---
layout: post
title: GSoc 2016
excerpt: "About my GSoC 2016 project"
modified: 2016-08-23
categories: articles
tags: [gsoc2016, github]
image:
  feature: gsoc.png
  credit: Google
  creditlink: http://summerofcode.withgoogle.com/
comments: true
share: true
---

This post is going to be about my GSoC 2016 project under Open Genome Informatics
organisation along with Anurag Priyam and Yannick Wurm as my mentors.

## About Project
Problem statement: BLAST visualizations library for Bioruby and SequenceServer.
Brief explanation: It is now trivial to generate large amounts of DNA sequence data; the challenge lies in making sense of the data. In many cases, researchers can gain important insights by comparing newly obtained sequences to previously known sequences using BLAST (>100,000 citations). This project will focus on creating a reusable visualizations library for BLAST results and integrating it with [SequenceServer](http://sequenceserver.com), a popular BLAST server.
BLAST results are text based but lack rich visual representation. Having a visualizations can greatly facilitate interpretation of data.

## Warming Up
Before the project has started I was fairly acquainted with Ruby and Javascript. So I started with small bug fixes in order to get acquainted with the SequenceServer code. SequenceServer provided support for download the report generated in XML or TSV format. When one clicked the download button, it would generate the files and store it as tmp file until it is completely downloaded. But this process was repeated every time download was clicked, so we decided to save the tmp files generated in two formats, so if user needs it again no need to generate the file instead directly start the download. I also played around with some error handling issues just to get comfortable with the ruby part of the project. I helped improving the XML parsing of files and check for integrity issues and cases when a specific report user was searching and not found. I added checks in various places and raised appropriate error messages to help user figure out what was going wrong.

## Visualizations
So by this time I had played around the SequenceServer code enough to know how it was working and it was time to get down to the real part of the project. I started up with Length Distribution graph. It is  a simple histogram representing hit sequences length frequency. The rectangle were colored using a grey scale, where darker the shade meant most significant hit. This graph provided user with an idea about the all the length of the hit sequence and the length of the query sequence. It will also help user in annotations, identify proteins across species, validate gene predictions, etc The graph was drawn using d3.js.
Next I started with Circos visualizations. Currently SequenceServer has Kablammo to show alignment mapping between hit sequence and query sequence, Alignment overview to show alignment mapping between query sequence and all its hit sequences. Now Circos visualizations will add alignment mapping of multiple query sequence and hit sequence, to its arsenal.  Circos visualizations is simple circos based graph with chords representing query sequence and hit sequence and ribbons represent the alignment mappings. The chords representing the query sequence is green in color and the others representing the hit sequence are blue in color. The ribbons are colored in red-yellow-blue color schema with red representing the most significant hit for a query and blue as least significant hit. One can hover over the ribbons to view its details such as the area this specific alignment covers on query or hit sequence, and the probability that this match was by chance. One can click on a  chord to view its alignments. This was drawn using CircosJs

Later I started with refinements of the previous graphs that SequenceServer provided. Since now we had four different visualizations and many of them using so much of common code we decided to make the code modular, in order to make the code look better and to facilitate adding of the new visualizations in future an much easier task  and also to make changes in current ones easily. In Kablammo the polygon were numbered alphabetical in order make it easier for user to understand which polygon corresponds to which alignment details provided below the graph.

My initial proposal was to add four new visualizations, but in detailed discussion with my mentors we decided that with the level of detailing required by one visualizations we should limit ourselves to two.

Here is my [list of commits](https://github.com/wurmlab/sequenceserver/commits/master?author=aniarya82)
